docs/madr/decisions/031-locality-aware-load-balancing.md

```diff
@@ -0,0 +1,479 @@
+              type: Only
+              zones: ["us-1", "us-2", "us-3"]
+          - from:
+              zones: ["eu-1", "eu-2", "eu-2"]
+            to:
+              type: Only
+              zones: ["eu-1", "eu-2", "eu-2"]
+```
+
+#### Load balance equally to local and zone
+
+![Use case 7](assets/031/use_case_7.png)
+
+We have the couple of use cases for this, but no config yet:
+- You are migrating from universal to k8s, both zones are in the same AZ and I would like to move traffic gradually from one to another
+- You have have some zone with testing or preview environment and you would like to handle live traffic on it for testing. Is it a use case for locality aware lb? Or should this be done with MeshHTTPRoute?
+The question is if we really need it, or can this be done currently in some other way?
+
+### Implementation
+
+We want to enhance current `MeshLoadBalancingStrategy` so code related to the changes should be placed in the plugin.
+
+#### Matching endpoints
+Each instance of the service might be in a different locality: zone, node. Users need to properly tag these dataplanes so we are able to match them. We need to retrieve all tags from all inbounds and, later, based on these tags, match outbounds that are in the same location.
+
+Example:
+
+On Kubernetes, all tags from the node could be populated automatically (see details later), while on Universal, users need to populate tags themselves:
+
+```yaml
+type: Dataplane
+mesh: default
+name: demo
+networking:
+  address: 127.0.0.1
+  inbound:
+    - port: 9999
+      servicePort: 10000
+      serviceAddress: 127.0.0.1
+      tags:
+        kuma.io/service: demo
+        kuma.io/protocol: tcp
+        k8s.io/az: us-east-1
+```
+
+
+Later, based on these tags, the control-plane creates `ClusterLoadAssignment` and puts all tags, for each endpoint, into cluster metadata::
+
+```json
+"metadata":{
+  "filter_metadata": {
+    "envoy.lb": {
+      "az/test": "b",
+      "kuma.io/protocol": "tcp",
+      "k8s.io/az": "us-east-1"
+    }
+  }
+}
+```
+
+
+We can retrieve them from `ClusterLoadAssignment` and match them with the policy. First, we need to retrieve all tags from inbounds so we can determine our location and match it with endpoints.
+
+Pseudo algorithm:
+
+1. Get all inbound tags.
+2. Take the matched rules.
+3. Check if inbound has the tag defined in the rule. 
+4. If yes, iterate over all endpoints of the service and group them in to priority groups, by matching with the specific rule(first localZone, then  crossZone). If no, get all the endpoints in the zone. 
+   1. Move matching endpoints by first tag to priority 0
+   2. Move matching endpoints by n tag to priority  n-1
+   3. All not matching endpoint in the local zone move to priority n
+   4. Iterate over zone endpoints and based on matching put them in priority N+ iteration
+5. Create `LocalityLbEndpoints` with selected endpoints. 
+6. Override `ClusterLoadAssignment` for the dataplane in the `ResourceSet`.
```

Comment:
Currently we do Mesh*Routes and then LaLB:

https://github.com/kumahq/kuma/blob/f3a3e973376967635fbc328948ce7b131ec7f71e/pkg/plugins/policies/policies.go#L21-L23

The routes create the right cluster with the right endpoints in it it seems right?
So sounds like what we need to do is take each cluster and reorganise the endpoints in groups with the right weights right?

This would be cacheable from within the policy I think (the key here would maybe be a little tricky but not impossible).

Reply:
In any case it really feels like the CLA could be only used in Routes.